core/server/worker/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -62,38 +70,102 @@
+  private static final ByteBuf CANCEL = Unpooled.buffer(0);
+  private static final ByteBuf ABORT = Unpooled.buffer(0);
+
   private ReentrantLock mLock = new ReentrantLock();
   /** The buffer for packets read from the channel. */
   @GuardedBy("mLock")
   private Queue<ByteBuf> mPackets = new LinkedList<>();
-  /** Set to true if the packet writer is active. */
+
+  /**
+   * Set to true if the packet writer is active.
+   *
+   * The following invariants (happens-before orders) must be maintained:
+   * 1. When mPacketWriterActive is true, it is guaranteed that mPackets is polled at least
+   *    once after the lock is released. This is guaranteed even when there is an exception
+   *    thrown when writing the packet.
+   * 2. When mPacketWriterActive is false, it is guaranteed that mPackets won't be polled before
+   *    before someone sets it to true again.
+   *
+   * The above are achieved by protecting it with "mLock". It is set to true when a new packet
+   * is read when it is false. It set to false when one of the these is true: 1) The mPackets queue
+   * is empty; 2) The write request is fulfilled (eof or cancel is received); 3) A failure occurs.
+   */
+  @GuardedBy("mLock")
+  private boolean mPacketWriterActive;
+
+  /**
+   * The error seen in either the netty I/O thread (e.g. failed to read from the network) or the
+   * packet writer thread (e.g. failed to write the packet).
+   */
   @GuardedBy("mLock")
-  private boolean mPacketWriterActive = false;
+  private Error mError;
 
-  protected volatile WriteRequestInternal mRequest = null;
+  private class Error {
+    final Throwable mCause;
+    final boolean mNotifyClient;
+    final Protocol.Status.Code mErrorCode;
 
-  protected abstract class WriteRequestInternal implements Closeable {
-    // This ID can either be block ID or temp UFS file ID.
-    public long mId = -1;
-    public long mSessionId = -1;
+    Error(Throwable cause, boolean notifyClient, Protocol.Status.Code code) {
+      mCause = cause;
+      mNotifyClient = notifyClient;
+      mErrorCode = code;
+    }
   }
 
   /**
-   * The next pos to queue to the buffer. Updated by the packet reader. Mostly used to validate
-   * the validity of the request.
+   * mRequest is initialized only once for a whole file or block in
+   * {@link DataServerReadHandler#channelRead(ChannelHandlerContext, Object)}.
+   * After that, it should only be used by the packet writer thread.
+   * It is safe to read those final primitive fields (e.g. mId, mSessionId) if mError is not set
+   * from any thread (not such usage in the code now). It is destroyed when the write request is
+   * done (complete or cancel) or an error is seen.
    */
-  private volatile long mPosToQueue = 0;
+  protected volatile WriteRequestInternal mRequest;
+
+  abstract class WriteRequestInternal implements Closeable {
+    /** This ID can either be block ID or temp UFS file ID. */
+    final long mId;
+    final long mSessionId;
```

Comment:
It looks like this can be negative -1 sometimes. Maybe that should be commented on, and/or add a constant for that?